packages/core/database/lib/query/helpers/streams/readable.js

```diff
@@ -0,0 +1,176 @@
+    const query = this._query;
+
+    // Remove the original offset & limit properties from the query
+    // Theoretically, they would be replaced by calling them again, but this is just to be sure
+    query.clear('limit').clear('offset');
+
+    // Define the maximum read size based on the limit and the requested size
+    // NOTE: size is equal to _batchSize by default. Since we want to allow customizing it on
+    // the fly, we need to use its value instead of batchSize when computing the maxReadSize value
+    const maxReadSize =
+      // if no limit is defined in the query, use the given size,
+      // otherwise, use the smallest value between the two
+      this._limit === null ? size : Math.min(size, this._limit);
+
+    // Compute the limit for the next query
+    const limit =
+      // If a limit is defined
+      this._limit !== null &&
+      // And reading `maxReadSize` would fetch too many entities (> _limit)
+      this._fetched + maxReadSize > this._limit
+        ? // Then adjust the limit so that it only get the remaining entities
+          this._limit - this._fetched
+        : // Else, use the max read size
+          maxReadSize;
+
+    // If we don't have anything left to read (_limit === _fetched),
+    // don't bother making the query and end the stream by pushing null
+    if (limit <= 0) {
+      this.push(null);
+      return;
+    }
+
+    // Compute the offset (base offset + number of entities already fetched)
+    const offset = this._offset + this._fetched;
+
+    // Update the query with the new values (offset + limit)
+    query.offset(offset).limit(limit);
+
+    // Lock the ._destroy()
+    this[knexPerformingQuery] = true;
+
+    let results;
+    let count;
+    let err;
+
+    try {
+      // Execute the query and store the results & count
+      results = await query;
+
+      const { populate } = this._qb.state;
+
+      // Apply populate if needed
+      if (populate && results) {
+        await applyPopulate(results, populate, { qb: this._qb, uid: this._uid, db: this._db });
+      }
+
+      // Map results if asked to
+      if (this._mapResults && results) {
+        results = fromRow(this._meta, results);
+      }
+
+      count = results.length;
+    } catch (e) {
+      err = e;
+    }
+
+    // Unlock the ._destroy()
+    this[knexPerformingQuery] = false;
+
+    // Tell ._destroy() that it's now safe to close the db connection
+    // Q: Should we push & return entities anyway or should we assume that the user wanted to abort the query?
+    if (this.destroyed) {
+      this.emit(knexQueryDone);
+      return;
+    }
+
+    // If there is an error, destroy with the given error
+    if (err) {
+      this.destroy(err);
+      return;
+    }
+
+    // Update the amount of fetched entities
+    this._fetched += count;
+
+    // While there is at least one value to unpack
+    while (results.length) {
+      // Add the first one to the stream's queue
+      this.push(results.shift());
+    }
```

Comment:
Unless I'm getting something wrong with the logic, I'd prefer:
```
    if (results.length) {
      results.forEach(this.push);
    }
```
to avoid a lot of potentially costly shifts.

Reply:
I initially did that to clear memory over time but it might not be worth regarding to the added time complexity, I'll add a loop instead